src/rust/cryptography-x509-validation/src/certificate.rs

```diff
@@ -0,0 +1,116 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+use asn1::ObjectIdentifier;
+use cryptography_x509::{
+    certificate::Certificate,
+    extensions::{Extension, Extensions},
+    oid::{
+        AUTHORITY_KEY_IDENTIFIER_OID, BASIC_CONSTRAINTS_OID, EXTENDED_KEY_USAGE_OID, KEY_USAGE_OID,
+        SUBJECT_ALTERNATIVE_NAME_OID, SUBJECT_KEY_IDENTIFIER_OID,
+    },
+};
+
+use crate::{backend::Backend, extensions::ExtensionError, types::Name};
+
+#[derive(Debug, PartialEq)]
+pub enum CertificateError {
+    DuplicateExtension(asn1::ObjectIdentifier),
+    InvalidExtension(ExtensionError),
+}
+
+impl From<ExtensionError> for CertificateError {
+    fn from(value: ExtensionError) -> Self {
+        CertificateError::InvalidExtension(value)
+    }
+}
+
+pub(crate) trait CertificateExt<'a>: private::Sealed {
```

Comment:
As a general proposition, I don't like extension traits. If these APIs make sense, we can just put them on Certificate, or they can be free-functions.

Reply:
Yeah, all of these should appear on `Certificate` IMO (and respectively same for the `Extension` extensional trait). They were only here originally for development velocity/so that I could keep modifications to a single crate at a time, but I think that's outlived its value.